elastic/kibana
# [Snapshot Restore] Fix logic for finding last successful managed snapshot

Fixes https://github.com/elastic/kibana/issues/158548
Fixes https://github.com/elastic/kibana/issues/153107

## Summary

This PR changes the logic for finding the last successful managed snapshot by searching amongst all snapshots rather than searching only from the filtered snapshots. This way, the last successful managed snapshot will always be the same, no matter whether the list is filtered or not.

The plugin README is updated with instructions on how to imitate a cloud-managed repository in order to create a managed snapshot.

Also, functional tests are added for checking the logic around last successful snapshots and filtering (the test suite successfully catches the incorrect behaviour when the fix from this PR is not present).

**How to test:**
We need to mock the cloud environment in order to have a managed repository, policy, and snapshots. Also, we add two different repo paths because we will create two repositories - one managed and one unmanaged so that we test a scenario with a list of managed and unmanaged snapshots.

1. Start Es with `yarn es snapshot --license trial -E path.repo='/tmp/es-backups','/tmp/snap'` and Kibana with `yarn start`
2. Add the `cluster.metadata.managed_repository` and `cluster.metadata.managed_policies` settings via Console:
```
PUT /_cluster/settings
{
  "persistent": {
    "cluster.metadata.managed_repository": "found-snapshots",
    "cluster.metadata.managed_policies": ["managed-policy"]
  }
}
```
4. Go to Stack Management -> Snapshot and Restore
5. Create a Shared File System repository  with the same name as your managed repository setting value (`found-snapshots`) and with location `/tmp/es-backups`
6. Create a second Shared File System repository (this one will be unmanaged) with the name `unmanaged-repo` and with location `/tmp/snap`
7. Create a managed policy with the same name as your managed policy setting value (`managed-policy`) that uses the managed repository created in step 5.
8. Create a second unmanaged policy that uses the unmanaged repository created in step 6.
9. Run the two policies a couple of times so that multiple snapshots are generated.
10. Go to the Snapshots tab and verify that the most recent managed snapshot is non-deletable.
11. Filter the snapshots so that you get another snapshot on the top row. Verify that this snapshot is deletable.

https://github.com/elastic/kibana/assets/59341489/9b25998b-875a-4a1d-a704-2a224766dd6f
